miniapp/src/test/java/com/rakuten/tech/mobile/miniapp/RealMiniAppSpec.kt

```diff
@@ -238,6 +246,73 @@ class RealMiniAppSpec {
 
         assertEquals(expected, actual)
     }
+    /** end region */
+
+    /** region: RealMiniApp.listDownloadedWithCustomPermissions */
+    @Test(expected = RequiredPermissionsNotGrantedException::class)
+    fun `verifyManifest will throw exception when required permissions are denied`() =
+        runBlockingTest {
+            val demoManifest = MiniAppManifest(
+                listOf(Pair(MiniAppCustomPermissionType.USER_NAME, "reason")), listOf(), mapOf()
+            )
+            val cachedManifest = CachedManifest(TEST_MA_VERSION_ID, demoManifest)
+            When calling downloadedManifestCache.readDownloadedManifest(TEST_MA_ID) itReturns cachedManifest
+
+            val permission = MiniAppCustomPermission(
+                TEST_MA_ID, listOf(Pair(MiniAppCustomPermissionType.USER_NAME, MiniAppCustomPermissionResult.DENIED))
+            )
+            When calling miniAppCustomPermissionCache.readPermissions(TEST_MA_ID) itReturns permission
+            When calling downloadedManifestCache.isRequiredPermissionDenied(permission) itReturns true
```

Comment:
It seems there's a lot of duplicate code in these three new tests. So I think you could either:

1. Pull some of the code into a separate function
2. OR pull this block into it's own test class (within this file) so you can have shared setup for these tests.

Actually, I think 2 is a good idea in general for `RealMiniAppSpec` - it would make the test organization much easier to read. For example:

```
@Ignore
class RealMiniAppSpec {
  // Common setup for the tests (doesn't contain any tests)
}

class RealMiniAppManifestSpec: RealMiniAppSpec() {
   @Before
   fun before() {
     // Shared setup for manifest tests
   }
}
```

Reply:
Yeah, I have updated as option 2.